triggers/daml/Daml/Trigger.daml

```diff
@@ -1,130 +1,164 @@
 -- Copyright (c) 2019 The DAML Authors. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
-{-# LANGUAGE AllowAmbiguousTypes #-}
+
 daml 1.2
 module Daml.Trigger
-  ( Message(..)
-  , Completion(..)
-  , CompletionStatus(..)
-  , Transaction(..)
-  , Identifier(..)
-  , AnyContractId(..)
-  , TransactionId(..)
-  , EventId(..)
-  , CommandId(..)
-  , Event(..)
-  , Created(..)
-  , Archived(..)
-  , Trigger(..)
-  , ActiveContracts(..)
-  , Commands(..)
-  , TemplateId(..)
-  , Command
-  , exerciseCmd
-  , createCmd
-  ) where
-
-data Identifier = Identifier
-  { packageId : Text
-  , moduleName : Text
-  , entityName : Text
-  } deriving (Show, Eq)
-
-data AnyContractId = AnyContractId
-  { templateId : Identifier
-  , contractId : Text
-  } deriving (Show, Eq)
-
-newtype TransactionId = TransactionId Text
-  deriving (Show, Eq)
-
-newtype EventId = EventId Text
-  deriving (Show, Eq)
-
-newtype CommandId = CommandId Text
-  deriving (Show, Eq)
-
-data Transaction = Transaction
- { transactionId : TransactionId
- , events : [Event]
- }
-
-data Event
-  = CreatedEvent Created
-  | ArchivedEvent Archived
-
-data Created = Created
-  { eventId : EventId
-  , contractId : AnyContractId
-  , argument : AnyTemplate
-  }
+ ( ACS
+ , getTemplates
+ , Trigger(..)
+ , TriggerA
+ , emitCommands
+ , runTrigger
+ , CommandId
+ , Command(..)
+ , AnyContractId(..)
+ , exerciseCmd
+ , createCmd
+ ) where
 
-data Archived = Archived
-  { eventId : EventId
-  , contractId : AnyContractId
-  } deriving (Show, Eq)
+import DA.Next.Map (Map)
+import qualified DA.Next.Map as Map
+import DA.Optional
 
-data Message
-  = MTransaction Transaction
-  | MCompletion Completion
+import Daml.Trigger.LowLevel hiding (Trigger)
+import qualified Daml.Trigger.LowLevel as LowLevel
 
-data Completion = Completion
-  { commandId : CommandId
-  , status : CompletionStatus
-  } deriving Show
+-- public API
 
-data CompletionStatus
-  = Failed { status : Int, message : Text }
-  | Succeeded { transactionId : TransactionId }
-  deriving Show
+newtype ACS = ACS [(AnyContractId, AnyTemplate)]
 
-data ActiveContracts = ActiveContracts { activeContracts : [Created] }
+getTemplates : Template a => ACS -> [(AnyContractId, a)]
+getTemplates (ACS tpls) = mapOptional (\(cid, tpl) -> (cid,) <$> fromAnyTemplate tpl) tpls
 
--- | Trigger is (approximately) a left-fold over `Message` with
--- an accumulator of type `s`.
 data Trigger s = Trigger
-  { initialState : Party -> ActiveContracts -> s
-  , update : Message -> s -> (s, [Commands], Text)
+  { initialize : ACS -> s
+  -- The ACS already includes the updates from the Message
+  , updateState : ACS -> Message -> s -> s
+  , rule : Party -> ACS -> Map CommandId [Command] -> s -> TriggerA ()
```

Comment:
Trying to understand these two functions. `updateState` is run on any `Message`, both `MCompletion` and `MTransaction`, but it can only determine the new state and not issue any new commands. `rule` is only run on `MTransaction`, and it has the ability to issue new commands. Is that correct? Why can't the user issue new commands after an `MCompletion`?

Reply:
In general, the idea is that `rule` is idempotent can basically be run at any point and should only depend on the ACS and the user state (the user state is often not even required, see the copy example in this PR). As mentioned in a comment, we do want to run the rule on failed completions but I’m leaving that for a separate PR. Running it on successful completions doesn’t make that much sense since in that case you will also get a transaction event and running the rule before the ACS has been updated can be a bit confusing.

We might want to change how and when exactly these events are triggered but I’d like to see some more examples before we finalize on something.

Reply:
Okay, thanks for explaining. Didn't see that comment.